src/core/mac/mac_frame.cpp

```diff
@@ -145,6 +145,148 @@ ThreadError Frame::InitMacHeader(uint16_t aFcf, uint8_t aSecurityControl)
     return kThreadError_None;
 }
 
+ThreadError Frame::ValidatePsdu(void)
+{
+    ThreadError error = kThreadError_Parse;
+    uint8_t offset = 0;
+    uint16_t fcf;
+    uint8_t footerLength = kFcsSize;
+
+    VerifyOrExit(GetPsduLength() > kFcfSize + kDsnSize, ;);
+
+    fcf = static_cast<uint16_t>((GetPsdu()[1] << 8) | GetPsdu()[0]);
+
+    offset += kFcfSize + kDsnSize;
+
+    // Destinatinon PAN + Address
+    switch (fcf & Frame::kFcfDstAddrMask)
+    {
+    case Frame::kFcfDstAddrNone:
+        break;
+
+    case Frame::kFcfDstAddrShort:
+        VerifyOrExit((uint8_t)(GetPsduLength() - offset) > sizeof(PanId) + sizeof(ShortAddress), ;);
+        offset += sizeof(PanId) + sizeof(ShortAddress);
+        break;
+
+    case Frame::kFcfDstAddrExt:
+        VerifyOrExit((uint8_t)(GetPsduLength() - offset) > sizeof(PanId) + sizeof(ExtAddress), ;);
+        offset += sizeof(PanId) + sizeof(ExtAddress);
+        break;
+
+    default:
+        goto exit;
+    }
+
+    // Source PAN + Address
+    switch (fcf & Frame::kFcfSrcAddrMask)
+    {
+    case Frame::kFcfSrcAddrNone:
+        break;
+
+    case Frame::kFcfSrcAddrShort:
+        if ((fcf & Frame::kFcfPanidCompression) == 0)
+        {
+            VerifyOrExit((uint8_t)(GetPsduLength() - offset) > sizeof(PanId), ;);
+            offset += sizeof(PanId);
+        }
+
+        VerifyOrExit((uint8_t)(GetPsduLength() - offset) > sizeof(ShortAddress), ;);
+        offset += sizeof(ShortAddress);
+        break;
+
+    case Frame::kFcfSrcAddrExt:
+        if ((fcf & Frame::kFcfPanidCompression) == 0)
+        {
+            VerifyOrExit((uint8_t)(GetPsduLength() - offset) > sizeof(PanId), ;);
+            offset += sizeof(PanId);
+        }
+
+        VerifyOrExit((uint8_t)(GetPsduLength() - offset) > sizeof(ExtAddress), ;);
+        offset += sizeof(ExtAddress);
+        break;
+
+    default:
+        goto exit;
+    }
+
+    // Security Header
+    if (fcf & Frame::kFcfSecurityEnabled)
+    {
+        VerifyOrExit((uint8_t)(GetPsduLength() - offset) > sizeof(uint8_t), ;);
+
+        uint8_t secControl = GetPsdu()[offset];
+        offset += sizeof(uint8_t);
+
+        if (secControl & kSecLevelMask)
```

Comment:
The Security Control and Frame Counter fields are present whenever the FCF Security Enabled flag is set.  Furthermore, this logic is double-counting the Security Control field.  I think you can just remove the `offset` increment on line 219 and remove this if-statement.  Keep the check on line 223 and increment on line 224.